15/umbraco-cms/tutorials/inspiration/migrating-macros/migrating-macros.md

```diff
@@ -0,0 +1,889 @@
+                Culture = null,
+                Segment = null,
+            });
+            // setup the layout
+            var layoutItem = new RichTextBlockLayoutItem(blockKey);
+            var layoutList = richTextEditorValue.Blocks.Layout.ContainsKey(Constants.PropertyEditors.Aliases.RichText)
+                ? richTextEditorValue.Blocks.Layout[Constants.PropertyEditors.Aliases.RichText].ToList()
+                : new List<IBlockLayoutItem>();
+            layoutList.Add(layoutItem);
+            richTextEditorValue.Blocks.Layout[Constants.PropertyEditors.Aliases.RichText] = layoutList;
+            
+            // now that we have converted the data into a block, replace the macro tag by a block tag
+            richTextEditorValue.Markup = richTextEditorValue.Markup.ReplaceFirst(macroMatch.Value, $"<umb-rte-block-inline data-content-key=\"{blockKey}\"></umb-rte-block-inline>");
+        }
+        
+        return _jsonSerializer.Serialize(richTextEditorValue);
+    }
+
+    private BlockPropertyValue CreateInvariantTextboxBlockPropertyValue(string alias, string value)
+        => new BlockPropertyValue
+        {
+            Alias = alias,
+            Culture = null,
+            Segment = null,
+            Value = value,
+            // The propertyType here is just a dummy to allow us to save the propertyEditorAlias
+            PropertyType = new PropertyType(_shortStringHelper,Constants.PropertyEditors.Aliases.TextBox, ValueStorageType.Nvarchar)
+        };
+}
+```
+
+# Retrieving the data
+In our setup, we will be retrieving the values straight from the database using custom dto's, this allows us to just get the data we need. This example does not take nested data into account. For an example on how to to do this, check out one of the alternatives at the bottom of this article.
+This example also only fetches the active (draft/current) version of the affected data to reduce processing time.
+
+# Putting the updated value back
+Once we have transformed the data we need to store it. One again we have chosen to use some custom sql to perform this. If you however need to perform validation on the updated value, you will either have to use some higher level services (`IContentValidationService`/`IContentEditingService`) or use the `RichTextPropertyValueEditor.Validate()` method. Because this examples fetches the current data and overwrites the value it, the old value will not show up in the version history of the affected node. If you do need this to happen then we advice you to use the `IContentValidationService` or `IContentService` instead.
+
+# Code time!
+Now that we have a converter we need a way to call the correct one depending on the macros found in an RTE value. For this we will Create a `MacroMigrationService` that holds the following method.
+
+```csharp
+public void Migrate(IEnumerable<int> PropertyDataIds)
+    {
+        using IScope scope = _scopeProvider.CreateScope(autoComplete: true);
```

Comment:
Do not use `autoComplete: true` when writing to the database, as exceptions withing the using block won't cause the transaction to be aborted/rolled back: remove this parameter and explicitly call `scope.Complete()` below.